File: backend/internal/media/airplay.go
package media

// AirPlay is the one route where HomeHub does not hand content to a speaker
// but pushes it at one. That inverts who owns the audio, so it needs a little
// more vocabulary than the other routes: a description of the receiver good
// enough to open a session with, a host that owns the sending, and a way to
// reach a receiver that is already being driven.
//
// None of the protocol lives here. This package knows that a receiver has an
// address and takes PCM or ALAC; internal/airplay knows what any of that means
// on the wire. That split is the same one StreamHost draws, and it is what
// keeps the route engine testable without a socket.

import "context"

// AirPlayDest is where and how to reach one receiver. Built by the endpoint
// adapter from what the receiver advertised, and passed through the host
// untouched.
type AirPlayDest struct {
	// ID is the endpoint id, which is also how a live cast is looked up
	// again when someone changes the volume mid-song.
	ID   string
	Name string
	Host string
	Port int
	// PCM and ALAC are the formats the receiver said it takes. Both are
	// bit-exact; which is used is the sender's choice.
	PCM  bool
	ALAC bool
	// NeedsEncryption is a receiver that will not accept cleartext audio.
	NeedsEncryption bool
	// Metadata is whether the receiver has a display to fill in.
	Metadata bool
	// Volume is the level to set before audio starts, 0-100, and zero means
	// "leave it where it is". Per receiver rather than per cast because an
	// AirPlay receiver keeps its own volume between senders: the level that
	// belongs to a kitchen box is not the one that belongs to a study box,
	// and a cast that levelled them together would undo whatever the
	// household had set.
	Volume int
}

// AirPlayTarget is CapAirPlay: the endpoint can be sent a pushed stream.
type AirPlayTarget interface {
	AirPlayDest() AirPlayDest
}

// AirPlayControl is a receiver inside a running cast.
//
// It exists because an AirPlay receiver has no state of its own to read or
// change: it is a sink, and everything about what it is playing lives in the
// sender. Volume is the exception — that is the receiver's own — and pause
// means "stop sending and drop what is buffered", which only the sender can
// do.
type AirPlayControl interface {
	SetVolume(ctx context.Context, level int) error
	Pause(ctx context.Context) error
	Resume(ctx context.Context) error
	Playing() bool
}

// AirPlayHost sends audio to receivers. Implemented by internal/airplay.
type AirPlayHost interface {
	// Cast pushes s to every destination and returns a stop function that
	// ends the cast and releases the receivers. All destinations or none:
	// a partial cast is silence in half a room with no error to explain it.
	Cast(ctx context.Context, s *Stream, dests []AirPlayDest) (stop func(), err error)
	// Live returns the control surface for an endpoint the running cast is
	// driving, or false when it is driving nothing.
	Live(id string) (AirPlayControl, bool)
}

File: frontend/src/lib/types/airplay.ts
// AirPlay receivers — RoPieee boxes, Apple TVs, anything running
// shairport-sync.
//
// A third speaker bridge beside Sonos and KEF, and the shape of it is thinner
// for a reason: those two are players that hold something, so their types
// carry queues, sources and settings. An AirPlay receiver is a *sink*. It has
// an address, a volume, and whatever HomeHub is currently sending it — there
// is no state on the device to read, so there is none to model here.
//
// Everything about playing to one is a zone operation through the media
// types, because pushing audio to receivers is a route, not a per-speaker
// command. See internal/airplay.

export interface AirPlaySpeaker {
  id: string;
  name: string;
  ip: string;
  /** The RTSP port from the receiver's advertisement. 7000 for most. */
  port: number;
  /** The receiver's own identity, normalised — the stable id across DHCP
   *  leases, the job `mac` does for KEF. Empty for a typed-in receiver. */
  device_id?: string;
  room?: string;
  model?: string;
  /** The formats the receiver advertised. Both are bit-exact, so this is
   *  about how the samples are packed, not about quality. */
  pcm: boolean;
  alac: boolean;
  /** A receiver that will not accept cleartext audio. */
  needs_encryption?: boolean;
  /** Whether the box also advertises AirPlay 2. Display only — HomeHub sends
   *  a classic session either way, and shairport-sync receivers (so every
   *  RoPieee) accept one whichever mode they are in. */
  airplay2?: boolean;
  /** Whether it has a display worth sending track info to. */
  metadata?: boolean;
  /** The level the household last set, 0-100. Stored because a receiver only
   *  takes a volume inside a session: this is what the next cast opens with. */
  volume?: number;
}

/** A registered receiver, as the status endpoint reports it. */
export interface AirPlaySpeakerView extends AirPlaySpeaker {
  /** Bridge-qualified id ("airplay:abc") — what a zone stores. */
  member: string;
  /** Whether HomeHub is sending to it right now. This is as close to
   *  "reachable" as a receiver gets: asking one whether it is there means
   *  opening a session, which would take it away from whatever else is
   *  playing to it. */
  casting: boolean;
}

/**
 * Whether a receiver takes the session HomeHub opens, as answered by the
 * receiver during the scan.
 *
 * Three states rather than a boolean, and the middle one is the point: an
 * AirPlay 2 advertisement does not say whether the box accepts a classic
 * sender, so the scan asks. `unknown` means nothing answered — the receiver
 * may simply be off — and is attempted rather than refused.
 */
export type AirPlayClassic = "yes" | "no" | "unknown";

/** One receiver found by a scan. */
export interface AirPlayCandidate {
  name: string;
  ip: string;
  port: number;
  id?: string;
  model?: string;
  version?: string;
  audio: { sample_rate: number; bit_depth: number; channels: number };
  needs_password: boolean;
  metadata: boolean;
  registered: boolean;
  /** Also speaks AirPlay 2. Not a blocker: see AirPlayClassic. */
  airplay2?: boolean;
  /** What the receiver itself said about the classic session. */
  classic: AirPlayClassic;
  /** Whether HomeHub could actually drive it. A scan that lists a
   *  FairPlay-only Apple TV as addable sets up a failure two taps later. */
  supported: boolean;
  /** Why not, in words meant for a person. Set when `supported` is false. */
  problem?: string;
}
